[not for land] further syncbn investigations

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40600 [not for land] further syncbn investigations**

Summary:

Context:

Not for land, just debugging.
Continuing to check what could be different between
NaiveSyncBN and BN.

Current state:

Running variance is biased in NaiveSyncBN and unbiased in SyncBN,
the difference is easily reproducible. Let's see if using biased
variance fixes the accuracy gap.

Test Plan:

https://www.internalfb.com/intern/paste/P136440735/

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22249330](https://our.internmc.facebook.com/intern/diff/D22249330)

vkuzo committed:
[wip] sync_bn: potentially fix accuracy issues

Summary:

Context:

There have been reported accuracy drops for SyncBatchNorm
in comparison to `NaiveSyncBatchNorm` (from Detectron2).  This PR's
goal is to investigate and ideally fix any numerical differences between
the two.  Note: both the code and this summary are a work in progress,
and a bunch of cleanup will be needed to land.

Current state:

Created a simple integration test to match numerics in forward and
backward passes.  This test uncovered that on master, numerics match with
relative differences of ~1e-16 to exact if `height * width` is a power of 2,
and there are relative differences of magnitude ~1e-8 to ~1e-9 otherwise.

From bisecting the NaiveSyncBatchNorm implementation, the difference
seems to be coming from calculating the local mean+var using
`torch.batch_norm_stats` compared to calculating it with `torch.mean`.

This could be due to either a bug in `torch.batch_norm_stats` or the
expected differences between its implementation compared to `torch.mean`.
the custom kernel is using two things:
* https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm
* https://developer.nvidia.com/blog/faster-parallel-reductions-kepler/

For now, let's run a few real world jobs (will take a few days) and see if
this change fixes the accuracy differences, to verify whether this
difference matters.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ngimel commented:
I thought accuracy issues with sync batch norm were resolved? 1e-8 - 1e-9 differences are to be expected due to floating point accuracy, it would be extremely surprising if they have a material effect on model accuracy.

vkuzo commented:
> I thought accuracy issues with sync batch norm were resolved? 1e-8 - 1e-9 differences are to be expected due to floating point accuracy, it would be extremely surprising if they have a material effect on model accuracy.

We are still seeing reports that SyncBN cannot match the accuracy of NaiveSyncBN for some large models such as EfficientNetB3.  I would also be very surprised if this particular PR matters, but it's the only difference I could find that's reproducible on a toy example. We will have data in a few days whether this matters or not.  If it's not this, we'll have to look into whether we can find any differences in real world settings.

vkuzo commented:
experiments showed that this doesn't matter

vkuzo committed:
Update on "[not for land] further syncbn investigations"


Summary:

Context:

Not for land, just debugging.
Continuing to check what could be different between
NaiveSyncBN and BN.

Current state:

Running variance is biased in NaiveSyncBN and unbiased in SyncBN,
the difference is easily reproducible. Let's see if using biased
variance fixes the accuracy gap.

Test Plan:

https://www.internalfb.com/intern/paste/P136440735/

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22249330](https://our.internmc.facebook.com/intern/diff/D22249330)

[ghstack-poisoned]

vkuzo committed:
Update on "[not for land] further syncbn investigations"


Summary:

Context:

Not for land, just debugging.
Continuing to check what could be different between
NaiveSyncBN and BN.

Current state:

Running variance is biased in NaiveSyncBN and unbiased in SyncBN,
the difference is easily reproducible. Let's see if using biased
variance fixes the accuracy gap.

Test Plan:

https://www.internalfb.com/intern/paste/P136440735/

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22249330](https://our.internmc.facebook.com/intern/diff/D22249330)

[ghstack-poisoned]

vkuzo committed:
Update on "[not for land] further syncbn investigations"


Summary:

Context:

Not for land, just debugging.
Continuing to check what could be different between
NaiveSyncBN and BN.

Current state:

Running variance is biased in NaiveSyncBN and unbiased in SyncBN,
the difference is easily reproducible. Let's see if using biased
variance fixes the accuracy gap.

Test Plan:

https://www.internalfb.com/intern/paste/P136440735/

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22249330](https://our.internmc.facebook.com/intern/diff/D22249330)

[ghstack-poisoned]
